common/src/primitives/height.rs

```diff
@@ -1,18 +1,113 @@
 use parity_scale_codec_derive::{Decode, Encode};
+use std::fmt;
+use std::ops::{Add, Sub};
+
+type HeightIntType = u64;
+type DistanceIntType = i64;
 
 #[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Encode, Decode)]
-pub struct BlockHeight(u64);
+pub struct BlockHeight(HeightIntType);
+
+// Display should be defined for thiserr crate
+impl fmt::Display for BlockHeight {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<BlockHeight> for HeightIntType {
+    fn from(block_height: BlockHeight) -> HeightIntType {
+        block_height.0
+    }
+}
+
+impl From<HeightIntType> for BlockHeight {
+    fn from(w: HeightIntType) -> BlockHeight {
+        BlockHeight(w)
+    }
+}
+
+impl Add<BlockDistance> for BlockHeight {
+    type Output = Option<BlockHeight>;
+
+    fn add(self, other: BlockDistance) -> Option<BlockHeight> {
+        let height: i64 = self.0.try_into().ok()?;
+        if height < 0 {
+            // we can't do arithmetic on this height anymore. Unless it's a bug, we won't face this in a million years
+            return None;
+        }
+
+        let result = height
+            .checked_add(other.0)
+            .expect("overflow when adding BlockHeight to instant");
```

Comment:
There seems to be a bit of inconsistency in error handling here. Just above, we check the block height fits into an `i64` but here, adding some distance to it is a hard error if the result is to large

Reply:
The idea here is that we should never be in a height of u64::MAX/2